### PR TITLE
[HxAutosuggest] HighlightFirstSuggestion #43

### PR DESCRIPTION
**Pull request by Harvey1214**

### CODE REVIEW

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/AutosuggestSettings.cs [hunk 1]
```diff
@@ -31,5 +31,10 @@ public record AutosuggestSettings : IInputSettingsWithSize
 		/// Input size.
 		/// </summary>
 		public InputSize? InputSize { get; set; }
+
+		/// <summary>
+		/// If true, the first suggestion is highlighted until another is chosen by the user.
+		/// </summary>
+		public bool? HighlightFirstSuggestion { get; set; } = true;
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/AutosuggestSettings.cs:38]: The default value cannot be set here.
The `Settings` has to be `null` to allow the evaluation cascade.
The default (true is correct) has to be set from `HxAutosuggest` static constructor (HxAutosuggest.nongeneric.cs).

File: Havit.Blazor.Components.Web.Bootstrap.Documentation/Shared/Components/Search/Search.razor [hunk 2]
```diff
@@ -1,19 +1,18 @@
...
-            <span class="p-2">No matches found</span>
-        </EmptyTemplate>
-    </HxAutosuggest>
-</form>
+<HxAutosuggest Placeholder="Search"
+                TItem="SearchItem"
+                TValue="SearchItem"
+                @bind-Value="@SelectedResult"
+                MinimumLength="1"
+                HighlightFirstSuggestion="true"
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap.Documentation/Shared/Components/Search/Search.razor:9]: `true` is default, no need to set here.

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/HxAutosuggest.cs [hunk 3]
```diff
@@ -169,6 +175,7 @@ protected override void BuildRenderInput(RenderTreeBuilder builder)
 			builder.AddAttribute(1023, nameof(HxAutosuggestInternal<TItem, TValue>.InputGroupStartTemplate), this.InputGroupStartTemplate);
 			builder.AddAttribute(1024, nameof(HxAutosuggestInternal<TItem, TValue>.InputGroupEndText), this.InputGroupEndText);
 			builder.AddAttribute(1025, nameof(HxAutosuggestInternal<TItem, TValue>.InputGroupEndTemplate), this.InputGroupEndTemplate);
+			builder.AddAttribute(1026, nameof(HxAutosuggestInternal<TItem, TValue>.HighlightFirstSuggestion), this.HighlightFirstSuggestion);
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/HxAutosuggest.cs:178]: You have to pass `HighlightFirstSuggestionEffective` here.
The `HxAutosuggestInternal` parameter name should be named `HighlightFirstSuggestionEffective` as well.

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestInternal.razor.cs [hunk 4]
```diff
@@ -98,6 +98,11 @@ public partial class HxAutosuggestInternal<TItem, TValue> : IAsyncDisposable
 		/// </summary>
 		[Parameter] public RenderFragment InputGroupEndTemplate { get; set; }
 
+		/// <summary>
+		/// If true, the first suggestion is highlighted until another is chosen by the user.
+		/// </summary>
+		[Parameter] public bool HighlightFirstSuggestion { get; set; }
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestInternal.razor.cs:104]: Rename to HighlightFirstSuggestionEffective please

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestInput.razor.cs [hunk 5]
```diff
@@ -38,6 +40,14 @@ private async Task HandleInput(ChangeEventArgs changeEventArgs)
 			await OnInputInput.InvokeAsync((string)changeEventArgs.Value);
 		}
 
+		private async Task HandleKeyDown(KeyboardEventArgs keyboardEventArgs)
+		{
+			if (keyboardEventArgs.Code == "Enter" || keyboardEventArgs.Code == "NumpadEnter")
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestInput.razor.cs:45]: Stylecop: Add parenthesis to primitive expressions.

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor [hunk 6]
```diff
@@ -4,14 +4,40 @@
...
+					type="button"
+					@onclick="async () => await HandleItemClick(currentItem)"
+					@onclick:stopPropagation="true"
+					tabindex="-1">
+
+					@ItemTemplate(currentItem)
+
+				</button>
+			}
+		}
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor:40]: No need to duplicate the code. If you set the attribute to `null`, it should remain unset. Something like this should be possible:
```
@onfocusin = "@(!HighlightFirstSuggestion ? null : () => hasFocus = true)"
```

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor.cs [hunk 7]
```diff
@@ -8,9 +8,34 @@ public partial class HxAutosuggestItems<TItem>
...
 
+		/// <summary>
+		/// Visually highlights the first suggestion.
+		/// </summary>
+		[Parameter] public bool HighlightFirstSuggestion { get; set; }
+
 		[Parameter] public RenderFragment EmptyTemplate { get; set; }
 		[Parameter] public string CssClass { get; set; }
 
+		private int hasFocusCount = 0;
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor.cs:19]: Why do we count the focuses when we are interested only in true/false? Is there any glitch behind that?

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor.cs [hunk 8]
```diff
@@ -8,9 +8,34 @@ public partial class HxAutosuggestItems<TItem>
 
 		[Parameter] public RenderFragment<TItem> ItemTemplate { get; set; }
 
+		/// <summary>
+		/// Visually highlights the first suggestion.
+		/// </summary>
+		[Parameter] public bool HighlightFirstSuggestion { get; set; }
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor.cs:14]: name HighlightFirstSuggestionEffective

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor.css [hunk 9]
```diff
@@ -0,0 +1,4 @@
+﻿::deep .hx-autosuggest-items-item-focus {
+    background-color: #e9ecef !important;
+    outline: -webkit-focus-ring-color auto 1px !important;
+}
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor.css:4]: Please consult with @crdo.
This seems to explicit to me, we should use CSS variable for color (preferably existing one from BS) etc.

**Review by hakenr — CHANGES_REQUESTED** [review]

File: Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor.css (hunk 9 above)
Comment (Harvey1214) [reply Havit.Blazor.Components.Web.Bootstrap/Forms/Autosuggests/Internal/HxAutosuggestItems.razor.css:4]: @crdo The color has been replaced with a Bootstrap CSS variable and the outline removed (to make identifying focused element clearer).

File: Havit.Blazor.Components.Web.Bootstrap.Documentation/Pages/Components/HxAutosuggestDoc/HxAutosuggest_Documentation.razor [hunk 10]
```diff
@@ -12,5 +12,8 @@
         <ComponentApiDocCssVariable Name="--hx-autosuggest-input-close-icon-opacity" Default=".25">
             Opacity of the close icon.
         </ComponentApiDocCssVariable>
+        <ComponentApiDocCssVariable Name="--hx-autosuggest-item-hightlighted-background-color" Default="var(--bs-gray-200)">
+            When <code>HxAutosuggest.HighlightFirstSuggestion</code> is <code>true</code>, then the first item receives this background color.
```
Comment (hakenr) [inline Havit.Blazor.Components.Web.Bootstrap.Documentation/Pages/Components/HxAutosuggestDoc/HxAutosuggest_Documentation.razor:16]: Is it just the first item, or does the CSS rule apply to all other items focused?

**Review by hakenr — CHANGES_REQUESTED** [review]

File: Havit.Blazor.Components.Web.Bootstrap.Documentation/Pages/Components/HxAutosuggestDoc/HxAutosuggest_Documentation.razor (hunk 10 above)
Comment (Harvey1214) [reply Havit.Blazor.Components.Web.Bootstrap.Documentation/Pages/Components/HxAutosuggestDoc/HxAutosuggest_Documentation.razor:16]: The class in "HxAutosuggestItems.razor.css" has been renamed to better convey the meaning.

**Pull request merged**